### ISSUE TITLE
Improve type assertion messages

### ISSUE THREAD

**Issue by rimutaka** [rustwasm/wasm-bindgen#3766]
### Motivation

Type checking functions inform what type was expected, but not what type was encountered. E.g. 

```javascript
function _assertBoolean(n) {
    if (typeof(n) !== 'boolean') {
        throw new Error('expected a boolean argument');
    }
}
```
It makes debugging harder because the encountered type can be anything and it requires manual edits to figure out what's going on. Time wasted.

### Proposed Solution

Change the message to something like

```javascript
        throw new Error(`expected a boolean argument, encountered ${typeof(n)}`);
```

### Alternatives

1. Manually edit _wasm_mod.js_ to print the encountered type for debugging
2. Print the type before the param reaches _wasm_mod.js_

Both feel wasteful. I think printing it out as part of the error has zero cost and saves a lot of grief.

It would also be inline with the helpful ways of Rust compiler.

### Would maintainers accept a PR for this?

I'm happy to work on a PR if there is interest to add this feature.

**Comment by daxpedda** [comment]
Yeah, I only see upsides.
Happy to review a PR!

**Comment by rimutaka** [comment]
@daxpedda , I hit a snag with adding tests.
The docs explain some of it, but a lot remains to be discovered by the contributors.
I'm slowly getting there :)

Would you consider a PR explaining the testing framework in more detail and some how-to's?

**Comment by daxpedda** [comment]
> Would you consider a PR explaining the testing framework in more detail and some how-to's?

I'm unsure exactly what you have in mind, but I'm happy to review any PRs improving documentation.
Just check up with me (or any other maintainer) before doing any significant amount of work.

**Comment by daxpedda** [comment]
Fixed in #3787.

**Comment by daxpedda** [comment]
Thanks @rimutaka, it's very appreciated!
